### PR TITLE
SetCurrentUser middleware

### PR DESCRIPTION
**Pull request by wakwak3125**
## WHAT
- Set current user to context by given access token

### CODE REVIEW

**Comment by wakwak3125** [comment]
@creasty 
please_cat

File: server/middleware/current_user.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+const (
+	currentuser = "CurrentUser"
+)
+
+// SetCurrentUser sets current authenticated user from authorization header
+func SetCurrentUser(appContext *system.AppContext) gin.HandlerFunc {
+	return func(c *gin.Context) {
+
```
Comment (creasty) [inline server/middleware/current_user.go]: Remove empty line

File: server/middleware/current_user.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"github.com/gin-gonic/gin"
+
+	oauth_service "github.com/torinos-io/api/service/oauth_service"
+	user_store "github.com/torinos-io/api/store/user_store"
+	"github.com/torinos-io/api/type/system"
+)
+
+const (
+	currentuser = "CurrentUser"
```
Comment (creasty) [inline server/middleware/current_user.go]: `currentUserContextName`

File: server/route/user.go [hunk 3]
```diff
@@ -8,8 +8,11 @@ import (
 
 // GetCurrentUser return the current user
 func GetCurrentUser(c *gin.Context) {
-	c.JSON(http.StatusOK, gin.H{
-		"message": "GetCurrentUser",
-		"user":    "yamada",
-	})
+	user, exists := c.Get("currentUser")
```
Comment (creasty) [inline server/route/user.go]: Don't use bare `c.Get("currentUser")`.
This leads to... well, IT IS happening now... It always becomes `exists = false` as you set with `CurrentUser` and get by `currentUser`.

Create `middleware.GetCurrentUser(c)`.

File: server/route/user.go [hunk 4]
```diff
@@ -8,8 +8,11 @@ import (
...
+
+	if exists {
+		c.JSON(http.StatusOK, user)
+	} else {
+		c.JSON(http.StatusBadRequest, nil)
```
Comment (creasty) [inline server/route/user.go]: Usually we use `401 Unauthorized`

File: service/oauth_service/find.go [hunk 5]
```diff
@@ -9,5 +9,10 @@ type FindRequest struct {
 
 // Find returns the found user
 func (s *service) Find(req *FindRequest) (*model.User, error) {
-	return s.UserStore.FindByGithubUser(req.UUID)
+	return s.UserStore.FindByGithubUUID(req.UUID)
+}
+
+// FindByAuthToken returns user that found by given access token
+func (s *service) FindByAuthToken(token string) (*model.User, error) {
+	return s.UserStore.FindByAuthToken(token)
```
Comment (creasty) [inline service/oauth_service/find.go]: Not correct.
Multiple access token can be issued for a single user and it's likely to be changed 
after time elapsed.

So it's non-sense to compare two access token to identify user.
You'll need to lookup GitHub's user id (UUID) by access token, and then find a user with that UUID.

**Review by creasty — CHANGES_REQUESTED** [review]
Reviewed

File: service/oauth_service/find.go [hunk 6]
```diff
@@ -9,5 +17,23 @@ type FindRequest struct {
...
-	return s.UserStore.FindByGithubUser(req.UUID)
+	return s.UserStore.FindByGithubUUID(req.UUID)
+}
+
+// FindByAuthToken returns user that found by given access token
+func (s *service) FindByAuthToken(token string) (*model.User, error) {
+	githubUser := &model.GithubUser{}
+
+	resp, _ := resty.
+		SetDebug(s.Config.Env == "development").
```
Comment (creasty) [inline service/oauth_service/find.go]: Use `s.Config.IsDevelopment()`

File: service/oauth_service/find.go [hunk 7]
```diff
@@ -9,5 +17,23 @@ type FindRequest struct {
 
 // Find returns the found user
 func (s *service) Find(req *FindRequest) (*model.User, error) {
-	return s.UserStore.FindByGithubUser(req.UUID)
+	return s.UserStore.FindByGithubUUID(req.UUID)
+}
+
+// FindByAuthToken returns user that found by given access token
+func (s *service) FindByAuthToken(token string) (*model.User, error) {
```
Comment (creasty) [inline service/oauth_service/find.go]: `FindByAccessToken`?

File: server/route/user.go [hunk 8]
```diff
@@ -4,12 +4,15 @@ import (
...
 // GetCurrentUser return the current user
 func GetCurrentUser(c *gin.Context) {
-	c.JSON(http.StatusOK, gin.H{
-		"message": "GetCurrentUser",
-		"user":    "yamada",
-	})
+	if user := middleware.GetCurrentUser(c); user != nil {
+		c.JSON(http.StatusOK, user)
+	} else {
+		c.JSON(http.StatusUnauthorized, "Can not find authorized current user")
```
Comment (creasty) [inline server/route/user.go]: Use `.AbortWithStatus` with early exit.

```go
user := middleware.GetCurrentUser(c)
if user == nil {
    c.AbortWithStatus(http.StatusUnauthorized)
    return
}

c.JSON(http.StatusOK, user)
```

File: server/middleware/current_user.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+		}
+
+		c.Set(currentUserContextName, user)
+		c.Next()
+	}
+}
+
+// GetCurrentUser returns current authenticated user from context
+func GetCurrentUser(c *gin.Context) *model.User {
+	v, exists := c.Get(currentUserContextName)
```
Comment (creasty) [inline server/middleware/current_user.go]: Simplify by:

```go
if v, exists := c.Get(currentUserContextName); exists {
    if user, ok := v.(*model.User); ok {
        return user
    }
}

return nil
```

**Review by creasty — COMMENTED** [review]
Small points

**Comment by wakwak3125** [comment]
@creasty Thank you 👍 
Could you check one more?

File: server/middleware/current_user.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+const (
+	currentUserContextName = "CurrentUser"
+)
+
+// SetCurrentUser sets current authenticated user from authorization header
+func SetCurrentUser(appContext *system.AppContext) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		h := c.GetHeader("Authorization")
```
Comment (creasty) [inline server/middleware/current_user.go]: Conventionally we use the form of `Authorization: Bearer xxxxxxxxx`.

**Review by creasty — COMMENTED** [review]
The rest looks fine 👍

File: server/middleware/current_user.go (hunk 10 above)
Comment (wakwak3125) [reply server/middleware/current_user.go]: OK!

File: server/middleware/current_user.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+		
+		if h == "" {
+			c.Next()
+			return
+		}
+		
+		splited := strings.Split(h, "Bearer")
```
Comment (wakwak3125) [inline server/middleware/current_user.go]: @creasty 
Is this correct approach to get Beare token?

Comment (creasty) [reply server/middleware/current_user.go]: Well, maybe not.. it leaves undesired space here.

```
"Bearer xxxxxxxxx" --> ["", " xxxxxxxxx"]
                             ^
```

Take a look at notifications for example.
https://github.com/wantedly/notifications/blob/ea6a316558508abc529d7e059185c4abd654c03f/server/middleware/current_user.go#L61

Comment (wakwak3125) [reply server/middleware/current_user.go]: @creasty Thank you very much.
I checked notifications and use that  🙏

**Review by creasty — APPROVED** [review]
LG 👍

**Pull request merged**